### ISSUE TITLE
Bug: [no-unsafe-return] For `async` function, `any` to `unknown` return are not allowed

### ISSUE THREAD

**Issue by ngoclong19** [typescript-eslint/typescript-eslint#5987]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&sourceType=module&code=GYVwdgxgLglg9mABMOcCMAKAzjMBzAGwFMB5AIwCsBlKAJ1zwC5Es6GBKZ8AazDgHckAbwCwAKESJaRKCFpIAUlRIA5AHQAHAIa0sRbA2LlqbfOwDciAPRXEAJRlykUAJ4aiyOLURLVmnXqIMFiIWmAuauIAvuLiWlgukMjg0PBIKHAATAb4RpQ09PjMrIV4nIgACrRwALbBRAA8PHyCAHyIohKh-FowUMgyEAAWGADkQ1BQGow2RAAeWjUaxGoQtVajFuKS0rLyPsrq2rr6OLmk+aZlljb2jvuu7p7evkcBHsGh4ZFiMWJAA&eslintrc=N4KABGBEBOD2sBdIC4wOgVwKYBpxQAcBDaAZy2hSgAEEBPArUgY2gEsCEBaJgGzYB2CAPTEyFSHgiQx5aAHlObWANJVQEaQTgArLMySoA2vk1QAdMIQsVAMzYBzcztIrIpsAF0pmyK4zQzFgAKgxYVJAAtrAAJhi84fgAvj4yvBgOgmrGHpC0YSzsnDyk-ELuEN74kFgAHghYAjHZYCZmNaWCCMjQ+rCRkY0xWDGSuQTpmQLI+YyFHNx8XcK9zP2DTSNj7RMZgjP0c6wLJWUiq+tDI1y9AI4YbOwCDlyHWFzMABb6ANaCDhUvCAkkA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkqyARhWANrg20AmCkANOx0joSRdAA88kAWAC6-QUVgBReMkmU8mAK7p5NSCU3Jo6sJp3sAviEtA

### Repro Code

```typescript
function foo1(singleObjString: string): unknown {
  return JSON.parse(singleObjString); // Return type for JSON.parse is any.
}

async function foo2(singleObjString: string): Promise<unknown> {
  await fetch('http://example.com/');
  return JSON.parse(singleObjString); // Return type for JSON.parse is any.
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  plugins: ['@typescript-eslint'],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:@typescript-eslint/recommended-requiring-type-checking',
  ],
};
```

### Expected Result

According to an example on https://typescript-eslint.io/rules/no-unsafe-return/, return `JSON.parse` as `any` to `unknown` type is allowed.

I expect there is no linting error for return statement of both functions.

### Actual Result

The normal function `foo1` works as expected.

About the async variant `foo2`.
There is an error 'Unsafe return of an `any` typed value' for the return statement.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.42.1` |
| `@typescript-eslint/parser`            | `5.42.1` |
| `TypeScript`                           | `4.8.4` |
| `ESLint`                               | `8.27.0` |
| `node`                                 | `18.12.1` |

**Comment by bradzacher** [comment]
Please provide a playground repro.

https://typescript-eslint.io/play/

**Comment by ngoclong19** [comment]
@bradzacher, I edited the first comment to add playground link.

Btw, is this playground working?

In my PC, eslint gives two `@typescript-eslint/no-unused-vars` warnings for functions `foo1` and `foo2`.
But the playground says 'All is ok!'.

**Comment by bradzacher** [comment]
The playground does not support recommended configs because it's intended for ***minimal reproductions***. Using a recommended set is not a minimal repro.

---

This is working as intended because it's unsafe.

You're looking at it from the perspective of "there is an `await` in the function body, so the  `return` is actually doing an assignment from `any` to `unknown`, so the rule shouldn't report".

This isn't a strictly correct thought. Because you may return a promise from an async function, you are really assigning `any` to `Promise<unknown>`, which is why the rule reports.
